### PR TITLE
WIP: fix(gatsby-source-contentful) No Contentful assets causes build failures

### LINKED ISSUE #15397: ERROR #85901  GRAPHQL | gatsby-source-contentful requires Media uploaded to work
## Description

When I try to link my gatsby project with my self -created Contentful space, it shows this issue all the time

But when I try to link my gatsby to a contentful spaces with I created with their templates or examples, it runs fine

But for self created spaces, it shows this error....

Any solutions?

##Error Message

ERROR #85901  GRAPHQL

There was an error in your GraphQL query:

Unknown type "ContentfulFixed". Did you mean "ContentfulSecondTry", "ContentfulContentType", or "ContentfulSecondTryEdge"?

### Environment
I was in the latest  gatsby-source-contentful and gatsby.

Any solutions?

### LINKED ISSUE #16455: Unknown type "ContentfulFixed". Did you mean "ContentfulBlogPost", "ContentfulContentType", or "ContentfulBlogPostEdge"?
##  ERROR #85901 GRAPHQL

I'm trying to fetch my blogs from contentful and it's giving me a weird error.
The only reason I am opening this issue is because the error told me to so.

This is what the error said: 
Error: The result of this StaticQuery could not be fetched.

"This is likely a bug in Gatsby and if refreshing the page does not fix it, please open an issue in https://github.com/gatsbyjs/gatsby/issues"

gatsby info output: 

  System:
    OS: Windows 10
    CPU: (8) x64 Intel(R) Core(TM) i7-3632QM CPU @ 2.20GHz
  Binaries:
    npm: 6.9.0 - C:\Program Files\nodejs\npm.CMD
  Languages:
    Python: 3.7.2
  Browsers:
    Edge: 44.18362.1.0
  npmPackages:
    gatsby: ^2.13.51 => 2.13.51
    gatsby-plugin-sass: ^2.1.4 => 2.1.4
    gatsby-plugin-sharp: ^2.2.10 => 2.2.10
    gatsby-remark-images: ^3.1.7 => 3.1.7
    gatsby-remark-relative-images: ^0.2.3 => 0.2.3
    gatsby-source-contentful: ^2.1.18 => 2.1.18
    gatsby-source-filesystem: ^2.1.9 => 2.1.9
    gatsby-transformer-remark: ^2.6.11 => 2.6.11

error UNHANDLED REJECTION

  Error: The system cannot find the path specified.

  - envinfo.js:1 Function.e.exports.sync
    [v12.0.0]/[gatsby-cli]/[envinfo]/dist/envinfo.js:1:4908

  - envinfo.js:1 Object.copySync
    [v12.0.0]/[gatsby-cli]/[envinfo]/dist/envinfo.js:1:66886

  - envinfo.js:1 Object.t.writeSync
    [v12.0.0]/[gatsby-cli]/[envinfo]/dist/envinfo.js:1:48715

  - envinfo.js:1 
    [v12.0.0]/[gatsby-cli]/[envinfo]/dist/envinfo.js:1:46872

  - envinfo.js:1 
    [v12.0.0]/[gatsby-cli]/[envinfo]/dist/envinfo.js:1:46887

  - task_queues.js:88 processTicksAndRejections
    internal/process/task_queues.js:88:5

### PR DESCRIPTION
**Pull request by joshuagagne**
## Description

Inlines an image to guarantee build has an image to infer the model around.  

Not the biggest fan of this solution.  

Also tried to get a schema typed before hand to guarantee the schema types existed, but that runs into problems where multiple schema types of the same name exist. 

This does not actually make the fragments pass when executed on inside a query, as this mock item doesn't actually exist in Contentful the resolved urls are not accurate.  I toyed with the idea of actually having a local gatsby image file to reference if no assets are found. 

## Related Issues

Fixes #16455 
Fixes #15397

### CODE REVIEW

File: packages/gatsby-source-contentful/src/gatsby-node.js [hunk 1]
```diff
@@ -146,6 +146,57 @@ exports.sourceNodes = async (
   newState[createSyncToken()] = nextSyncToken
   setPluginStatus(newState)
 
+  // Guarantee at least one asset exists.
+  // The subsequent GraphQL fragments added by fragmeent.js have an infered model to build against.
+  // With no assets the fragment will cause an error
+  if (assets.length === 0) {
```
Comment (joshuagagne) [inline packages/gatsby-source-contentful/src/gatsby-node.js]: I don't believe this check is sufficient, as existingNodes aren't checked.  The mock asset would end up available to query in real data sets when the sync detects no changes.

**Comment by pieh** [comment]
> Also tried to get a schema typed before hand to guarantee the schema types existed, but that runs into problems where multiple schema types of the same name exist.

You mean when user defines the type? IMO this should be the way to go here instead of creating "mock" asset node, that might mean making fixes (or changes) in gatsby core to fix the issue you are talking about. Can you clarify what kind of issues you saw?

**Comment by joshuagagne** [comment]
Well specifically we can't add it with the `createSchemaCustomization`'s `createType` API  as that results in the type being defined twice when content _does_ exist.  

`onPreExtractQueries` where we used to filter out these fragments and check for cache existence could be a useful spot to check things. But the side effect to fix this would have to be moving / deleting the `fragments.js` files as the `query-compiler.js` finds it and pulls the fragements out.  I'm unfamiliar with the fragment parsing so I didn't go down this route, but you _could_ put a function inside fragements.js that could be executed to get a boolean on whether to skip processing of that file.  Seems inelegant.

In `sourceNodes` providing the default asset allows us to get into `normalize.js`'s `createAssetNodes` function and create a node.   Looks like this function is responsible for creating the ContentfulAsset type.

**Comment by joshuagagne** [comment]
I updated the code's default asset creation to instead call createNode directly.  On the positive side this looks much cleaner, on the down side it causes a warning to be logged because I didn't actually add an item. 

`warn The asset with id: defaultMediaItem, contains no file.`

Still, at least this would allow a user with a brand new contentful space to have a build pass immediately after adding the plugin.  

Is there a spot after this that would create schema definitions? `createNode` doesn't seem like it should.

**Comment by vladar** [comment]
> Well specifically we can't add it with the createSchemaCustomization's createType API as that results in the type being defined twice when content does exist.

This sounds wrong. Can you clarify a bit what do you mean by this? Do you get an error or some kind of warning?

One of the reasons why schema customization API was added in the first place is to fix this issue with missing content. But it shouldn't break at all when content exists.

**Comment by joshuagagne** [comment]
If you attempt to create the type in the `createSchemaCustomization` stage it results in the error:

```
UNHANDLED REJECTION Schema must contain uniquely named types but contains multiple types named "ContentfulFixed".
```

As far as I can tell defining the adding the type early clashes with the createNode later on.  I was hoping it'd recognize the type and merge the schemas. 

An alternate fix seems to be to just remove the graphQL rule that validates the fragments.  The fragments get created, and if used in a query will cause the same error message to be thrown, it'll just happen at run time instead of build time.  

Would I be correct that the unfortunate side effect of removing this rule is that `query-compiler.js` is not limited to only fragments but all user defined queries? So a custom graphQL query in a page / component that should fail because of a typeName miss match would not.

**Comment by pvdz** [comment]
Hi @joshuagagne . It looks like this PR is still a wip (as also suggested by the title). I'm going to put it in Draft mode. This won't change much for you. You can pull it out of draft mode at any time when you feel the PR is ready and people that were involved in the PR will still see the updates and everything. Thanks!

**Pull request closed without merging**